daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -2829,13 +2829,19 @@ sealed abstract class LengthExprBase(
   ) {
 
   protected final def leafReferencedElements = {
-    val arg = args(0).asInstanceOf[PathExpression]
+    val arg = args(0) match {
+      case arg: PathExpression => arg
+      case _ => Assert.invariantFailed("NodeInfo.Exists, but arg was not a PathExpression.")
+    }
+    // test_lengthPatternBinaryPatternLimit in section 12 shows that the length can
+    // be evaluated when it is an array of hex binary values
+    if (arg.inherentType != NodeInfo.HexBinary && arg.isPathToOneWholeArray)
```

Comment:
See comment in the source above. Making this permissive for hexbinary values was needed for that test to pass.

Reply:
This indicates a more significant issue.

The expresion "." when it corresponds to an element declaration that is an array, it does NOT denote the array, it denotes the one infoset node only. 

So if the array element is name 'a', then "." is not equivalent to "../a" which references the whole array. 

Furthermore, suppose you write ".[1]", and '.' corresonds to an element that is  declared to be an array. 

This was a subject of recent DFDL workgroup discussion, and the conclusion is that "." is never an array. So ".[1]" is always equivalent to just "." and refers to a single node only. 

So I think the path expression "." (A PathExpression containing one step: Self) is going to need special treatment.